File: v2.x/Examples/SciChartDemo/SciChartDemo/ExampleViews/OscilloscopeChartView.m
//
//  OscilloscopeChartView.m
//  SciChartDemo
//
//  Created by Yaroslav Pelyukh on 4/6/17.
//  Copyright © 2017 ABT. All rights reserved.
//

#import "OscilloscopeChartView.h"
#import <SciChart/SciChart.h>
#include <math.h>
#include "DataManager.h"
#include "OscilloscopePanel.h"

@implementation OscilloscopeChartView{
    DataSourceEnum _selectedSource;
    BOOL _isDigitalLine;
    
    SCIXyDataSeries * _dataSeries1;
    SCIXyDataSeries * _dataSeries2;
    
    SCIFastLineRenderableSeries * _rSeries;
    
    NSTimeInterval _timeInterval;
    NSTimeInterval _lastTimeDraw;
    CADisplayLink * _displayLink;
    
    double _phase0;
    double _phase1;
    double _phaseIncrement;
    
    id<SCIAxis2DProtocol> _xAxis;
    id<SCIAxis2DProtocol> _yAxis;
}

@synthesize surface;

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    
    if (self) {
        surface = [SCIChartSurface new];
        surface.translatesAutoresizingMaskIntoConstraints = NO;
        
        OscilloscopePanel * panel = (OscilloscopePanel *)[[NSBundle mainBundle] loadNibNamed:@"OscilloscopePanel" owner:nil options:nil]. firstObject;
        panel.translatesAutoresizingMaskIntoConstraints = NO;
        
        __weak typeof(self) wSelf = self;
        
        panel.seriesTypeTouched = ^(UIButton *sender) { [wSelf changeSeriesType:sender]; };
        panel.rotateTouched = ^(UIButton *sender) { [wSelf rotateChart:sender]; };
        panel.flippedVerticallyTouched = ^(UIButton *sender) { [wSelf flipVerticallyChart:sender]; };
        panel.flippedHorizontallyTouched = ^(UIButton *sender) { [wSelf flipHorizontallyChart:sender]; };
        
        [self addSubview:surface];
        [self addSubview:panel];
        
        NSDictionary *layout = @{@"SciChart1": surface, @"Panel": panel};
        
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-(0)-[SciChart1]-(0)-|" options:0 metrics:0 views:layout]];
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-(0)-[Panel]-(0)-|" options:0 metrics:0 views:layout]];
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(0)-[Panel(43)]-(0)-[SciChart1]-(0)-|" options:0 metrics:0 views:layout]];
        
        _phaseIncrement = M_PI * 0.1;
        _selectedSource = FourierSeries;
        
        [self initalizeSurfaceData];
    }
    return self;
}

-(void) changeSeriesType:(UIButton *)sender{
    UIAlertController * alertController = [UIAlertController alertControllerWithTitle:@"Data Source" message:@"Select data source or make the line Digital" preferredStyle:UIAlertControllerStyleActionSheet];
    UIAlertAction * action = [UIAlertAction actionWithTitle:@"Fourier" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        _selectedSource = FourierSeries;
        [surface.xAxes itemAt:0].visibleRange = [[SCIDoubleRange alloc]initWithMin:SCIGeneric(2.5) Max:SCIGeneric(4.5)];
        [surface.yAxes itemAt:0].visibleRange = [[SCIDoubleRange alloc]initWithMin:SCIGeneric(-12.5) Max:SCIGeneric(12.5)];
    }];
    [alertController addAction:action];

    action = [UIAlertAction actionWithTitle:@"Lisajous" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        _selectedSource = Lissajous;
        [surface.xAxes itemAt:0].visibleRange = [[SCIDoubleRange alloc]initWithMin:SCIGeneric(-1.2) Max:SCIGeneric(1.2)];
        [surface.yAxes itemAt:0].visibleRange = [[SCIDoubleRange alloc]initWithMin:SCIGeneric(-1.2) Max:SCIGeneric(1.2)];
    }];
    [alertController addAction:action];
    
    action = [UIAlertAction actionWithTitle:@"Make line Digital" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        _isDigitalLine = !_isDigitalLine;
        _rSeries.isDigitalLine = _isDigitalLine;
    }];
    [alertController addAction:action];

    alertController.popoverPresentationController.sourceRect = sender.bounds;
    alertController.popoverPresentationController.sourceView = sender;
    
    [[self currentTopViewController] presentViewController:alertController animated:YES completion:nil];
}

- (void)rotateChart:(UIButton *)sender {
    int xAlignment = [surface.xAxes itemAt:0].axisAlignment;
    if (++xAlignment > 4) {
        xAlignment = 1;
    }
    [surface.xAxes itemAt:0].axisAlignment = (SCIAxisAlignment)xAlignment;
    int yAlignment = [surface.yAxes itemAt:0].axisAlignment;
    if (++yAlignment > 4) {
        yAlignment = 1;
    }
    [surface.yAxes itemAt:0].axisAlignment = (SCIAxisAlignment)yAlignment;
}

- (void)flipVerticallyChart:(UIButton *)sender {
    BOOL flip = [surface.yAxes itemAt:0].flipCoordinates;
    [surface.yAxes itemAt:0].flipCoordinates = !flip;
}

- (void) flipHorizontallyChart:(UIButton *)sender {
    BOOL flip = [surface.xAxes itemAt:0].flipCoordinates;
    [surface.xAxes itemAt:0].flipCoordinates = !flip;
}

- (UIViewController *)currentTopViewController {
    UIViewController * topVC = UIApplication.sharedApplication.delegate.window.rootViewController;
    while (topVC.presentedViewController) {
        topVC = topVC.presentedViewController;
    }
    return topVC;
}

- (void)willMoveToWindow:(UIWindow *)newWindow {
    [super willMoveToWindow: newWindow];
    if(_displayLink == nil) {
        _timeInterval = 0.2;
        _lastTimeDraw = 0.0;
        _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(updateOscilloscopeData:)];
        [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
    } else {
        [_displayLink invalidate];
        _displayLink = nil;
    }
}

- (void)initalizeSurfaceData {
    _xAxis = [SCINumericAxis new];
    _xAxis.autoRange = SCIAutoRange_Never;
    _xAxis.axisTitle = @"Time (ms)";
    _xAxis.visibleRange = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(2.5) Max:SCIGeneric(4.5)];
    
    _yAxis = [SCINumericAxis new];
    _yAxis.autoRange = SCIAutoRange_Never;
    _yAxis.axisTitle = @"Voltage (mV)";
    _yAxis.visibleRange = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(-12.5) Max:SCIGeneric(12.5)];
    
    _dataSeries1 = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Double YType:SCIDataType_Double];
    _dataSeries1.acceptUnsortedData = YES;
    _dataSeries2 = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Double YType:SCIDataType_Double];
    _dataSeries2.acceptUnsortedData = YES;
    
    _rSeries = [SCIFastLineRenderableSeries new];
    _rSeries.isDigitalLine = _isDigitalLine;
    
    [surface.xAxes add:_xAxis];
    [surface.yAxes add:_yAxis];
    [surface.renderableSeries add:_rSeries];
    
    [SCIThemeManager applyDefaultThemeToThemeable:surface];
}

- (void)updateOscilloscopeData:(CADisplayLink*)displayLink {
    _lastTimeDraw = _displayLink.timestamp;
    
    DoubleSeries * doubleSeries = [DoubleSeries new];
    if (_selectedSource == Lissajous) {
        [DataManager setLissajousCurve:doubleSeries alpha:0.12 beta:_phase1 delta:_phase0 count:2500];
        [_dataSeries1 clear];
        [_dataSeries1 appendRangeX:doubleSeries.xValues Y:doubleSeries.yValues Count:doubleSeries.size];
        [_rSeries setDataSeries:_dataSeries1];
    } else {
        [DataManager setFourierSeries:doubleSeries amplitude:2.0 phaseShift:_phase0 count:1000];
        [_dataSeries2 clear];
        [_dataSeries2 appendRangeX:doubleSeries.xValues Y:doubleSeries.yValues Count:doubleSeries.size];
        [_rSeries setDataSeries:_dataSeries2];
    }
    
    _phase0 += _phaseIncrement;
    _phase1 += _phaseIncrement * 0.005;
}

@end

File: Examples/SciChartDemo/Examples/Examples2D/FeaturedApps/Oscilloscope/OscilloscopeChartView.swift
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2019. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// OscilloscopeChartView.swift is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

enum DataSource {
    case Fourier
    case Lisajous
}

class OscilloscopeChartView: OscilloscopeLayout {
    
    var _selectedSource: DataSource = .Fourier
    var _isDigitalLine = false
    
    let _dataSeries1 = SCIXyDataSeries(xType: .double, yType: .double)
    let _dataSeries2 = SCIXyDataSeries(xType: .double, yType: .double)
    
    let _rSeries = SCIFastLineRenderableSeries()
    
    var _lastTimeDraw: TimeInterval = 0
    var _displayLink: CADisplayLink!
    
    var _phase0 = 0.0
    var _phase1 = 0.0
    var _phaseIncrement = .pi * 0.1
    
    let _xAxis = SCINumericAxis()
    let _yAxis = SCINumericAxis()
    
    override func commonInit() {
        weak var wSelf = self;
        self.seriesTypeTouched = { wSelf?.changeSeriesType() }
        self.rotateTouched = { wSelf?.rotateChart() }
        self.flippedVerticallyTouched = { wSelf?.flipChartVertically() }
        self.flippedHorizontallyTouched = { wSelf?.flipChartHorizontally() }
    }
    
    override func initExample() {
        _xAxis.autoRange = .never
        _xAxis.axisTitle = "Time (ms)"
        _xAxis.visibleRange = SCIDoubleRange(min: 2.5, max: 4.5)
        
        _yAxis.autoRange = .never
        _yAxis.axisTitle = "Voltage (mV)"
        _yAxis.visibleRange = SCIDoubleRange(min: -12.5, max: 12.5)

        _dataSeries1.acceptsUnsortedData = true
        _dataSeries2.acceptsUnsortedData = true
        
        _rSeries.isDigitalLine = _isDigitalLine
        
        surface.xAxes.add(_xAxis)
        surface.yAxes.add(_yAxis)
        surface.renderableSeries.add(_rSeries)
    }
    
    @objc fileprivate func updateOscilloscopeData(_ displayLink: CADisplayLink) {
        _lastTimeDraw = _displayLink.timestamp
        
        let doubleSeries = SCDDoubleSeries()
        if (_selectedSource == .Lisajous) {
            SCDDataManager.setLissajousCurve(doubleSeries, alpha: 0.12, beta: _phase1, delta: _phase0, count: 2500)
            _dataSeries1.clear()
            _dataSeries1.append(x: doubleSeries.xValues, y: doubleSeries.yValues)
            _rSeries.dataSeries = _dataSeries1
        } else {
            SCDDataManager.setFourierSeries(doubleSeries, amplitude: 2.0, phaseShift: _phase0, count: 1000)
            _dataSeries2.clear()
            _dataSeries2.append(x: doubleSeries.xValues, y: doubleSeries.yValues)
            _rSeries.dataSeries = _dataSeries2
        }
        
        _phase0 += _phaseIncrement
        _phase1 += _phaseIncrement * 0.005
    }
    
    fileprivate func changeSeriesType() {
        let alertController = UIAlertController(title: "Data Source", message: "Select data source or make the line Digital", preferredStyle: .actionSheet)
        alertController.addAction(UIAlertAction(title: "Fourier", style: .default, handler: {[unowned self] (action: UIAlertAction) -> Void in
            self._selectedSource = .Fourier
            self.surface.xAxes[0].visibleRange = SCIDoubleRange(min: 2.5, max: 4.5)
            self.surface.yAxes[0].visibleRange = SCIDoubleRange(min: -12.5, max: 12.5)
        }))
        alertController.addAction(UIAlertAction(title: "Lisajous", style: .default, handler: {[unowned self] (action: UIAlertAction) -> Void in
            self._selectedSource = .Lisajous
            self.surface.xAxes[0].visibleRange = SCIDoubleRange(min: -1.2, max: 1.2)
            self.surface.yAxes[0].visibleRange = SCIDoubleRange(min: -1.2, max: 1.2)
        }))
        alertController.addAction(UIAlertAction(title: "Make line Digital", style: .default, handler: {[unowned self] (action: UIAlertAction) -> Void in
            self._isDigitalLine = !self._isDigitalLine
            self._rSeries.isDigitalLine = self._isDigitalLine
        }))
        
        var topVC = UIApplication.shared.delegate?.window??.rootViewController
        while ((topVC?.presentedViewController) != nil) {
            topVC = topVC?.presentedViewController
        }
        topVC?.present(alertController, animated: true, completion: nil)
    }
    
    fileprivate func rotateChart() {
        var xAlignment: Int = Int(surface.xAxes[0].axisAlignment.rawValue)
        xAlignment += 1
        if (xAlignment > 4) {
            xAlignment = 1;
        }
        surface.xAxes[0].axisAlignment = SCIAxisAlignment(rawValue: UInt(xAlignment))!
        
        var yAlignment: Int = Int(surface.yAxes[0].axisAlignment.rawValue)
        yAlignment += 1
        if (yAlignment > 4) {
            yAlignment = 1;
        }
        surface.yAxes[0].axisAlignment = SCIAxisAlignment(rawValue: UInt(yAlignment))!
    }
    
    func flipChartVertically() {
        let flip = surface.yAxes[0].flipCoordinates
        surface.yAxes[0].flipCoordinates = !flip
    }
    
    func flipChartHorizontally() {
        let flip = surface.xAxes[0].flipCoordinates
        surface.xAxes[0].flipCoordinates = !flip
    }
    
    override func willMove(toWindow newWindow: UIWindow?) {
        super.willMove(toWindow: newWindow)
        if(_displayLink == nil) {
            _lastTimeDraw = 0.0
            _displayLink = CADisplayLink.init(target: self, selector: #selector(updateOscilloscopeData))
            _displayLink.add(to: RunLoop.main, forMode: RunLoop.Mode.default)
        } else {
            _displayLink.invalidate()
            _displayLink = nil;
        }
    }
}
